controllers/metal3.io/hostfirmwaresettings_controller.go

```diff
@@ -0,0 +1,460 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	metal3v1alpha1 "github.com/metal3-io/baremetal-operator/apis/metal3.io/v1alpha1"
+	"github.com/metal3-io/baremetal-operator/pkg/provisioner"
+)
+
+// HostFirmwareSettingsReconciler reconciles a HostFirmwareSettings object
+type HostFirmwareSettingsReconciler struct {
+	client.Client
+	Log                logr.Logger
+	ProvisionerFactory provisioner.Factory
+}
+
+type rInfo struct {
+	log    logr.Logger
+	hfs    *metal3v1alpha1.HostFirmwareSettings
+	bmh    *metal3v1alpha1.BareMetalHost
+	events []corev1.Event
+}
+
+func (info *rInfo) publishEvent(reason, message string) {
+
+	t := metav1.Now()
+	hfsEvent := corev1.Event{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: reason + "-",
+			Namespace:    info.hfs.ObjectMeta.Namespace,
+		},
+		InvolvedObject: corev1.ObjectReference{
+			Kind:       "HostFirmwareSettings",
+			Namespace:  info.hfs.Namespace,
+			Name:       info.hfs.Name,
+			UID:        info.hfs.UID,
+			APIVersion: metal3v1alpha1.GroupVersion.String(),
+		},
+		Reason:  reason,
+		Message: message,
+		Source: corev1.EventSource{
+			Component: "metal3-hostfirmwaresettings-controller",
+		},
+		FirstTimestamp:      t,
+		LastTimestamp:       t,
+		Count:               1,
+		Type:                corev1.EventTypeNormal,
+		ReportingController: "metal3.io/hostfirmwaresettings-controller",
+	}
+
+	info.events = append(info.events, hfsEvent)
+}
+
+//+kubebuilder:rbac:groups=metal3.io,resources=hostfirmwaresettings,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=metal3.io,resources=hostfirmwaresettings/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=metal3.io,resources=firmwareschemas,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=metal3.io,resources=firmwareschemas/status,verbs=get;update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.6.4/pkg/reconcile
+func (r *HostFirmwareSettingsReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, err error) {
+
+	reqLogger := r.Log.WithValues("hostfirmwaresettings", req.NamespacedName)
+	reqLogger.Info("start")
+
+	// Get the corresponding baremetalhost in this namespace, if one doesn't exist don't continue processing
+	bmh := &metal3v1alpha1.BareMetalHost{}
```

Comment:
I was able to duplicate some situations in which the provisioning id changed but the bmh and hfs resources stayed up when starting localbmo and the pods were recreated. It may be contrived using that script but I think its safer to get the bmh resource each time as the performance hit is quite small. In addition there may be cases where the bmh has been deleted but the hfs resource hasn't been removed yet.